internal/impl/postgresql/pglogicalstream/logical_stream.go

```diff
@@ -169,98 +159,119 @@ func NewPgStream(ctx context.Context, config *Config) (*Stream, error) {
+		if outputPlugin != decodingPlugin {
+			return nil, fmt.Errorf("replication slot %s already exists with different output plugin: %s", config.ReplicationSlotName, outputPlugin)
 		}
-	})
-
-	stream.logger.Debugf("starting stream from LSN %s with snapshot name %s", lsnrestart.String(), stream.snapshotName)
-	// TODO(le-vlad): if snapshot processing is restarted we will just skip right to streaming...
-	if !freshlyCreatedSlot || !config.StreamOldData {
-		if err = stream.startLr(ctx, lsnrestart); err != nil {
+		if confirmedLSNFromDB > 0 {
+			stream.ackedLSN = confirmedLSNFromDB - 1
+		}
+		stream.logger.Debugf("starting stream from LSN %s", confirmedLSNFromDB.String())
+		if err = stream.startLr(ctx, confirmedLSNFromDB); err != nil {
 			return nil, err
 		}
-
 		go func() {
 			defer stream.shutSig.TriggerHasStopped()
-			if err := stream.streamMessages(lsnrestart); err != nil {
+			if err := stream.streamMessages(confirmedLSNFromDB); err != nil {
 				stream.errors <- fmt.Errorf("logical replication stream error: %w", err)
 			}
 		}()
-	} else {
-		go func() {
-			defer stream.shutSig.TriggerHasStopped()
-			if err := stream.processSnapshot(); err != nil {
+		cleanups = nil
+		return stream, nil
+	}
+
+	var snapshotter *Snapshotter
+	if config.StreamOldData {
+		// Create a temporary replication slot that just creates a snapshot and freezes the LSN for the snapshot.
+		// We make this temporary so that if the snapshotting phase fails, we restart the snapshotting phase
+		// instead of resuming from the start of the stream (with an incomplete snapshot).
+		_, snapshotName, err := CreateReplicationSlot(
+			ctx, stream.pgConn,
+			stream.slotName+"_tmp",
+			decodingPlugin,
+			CreateReplicationSlotOptions{
+				Temporary:      true,
+				SnapshotAction: "EXPORT_SNAPSHOT",
+			},
+		)
+		if err != nil {
+			return nil, fmt.Errorf("unable to create temporary replication slot for snapshot: %w", err)
+		}
+		stream.logger.Tracef("exported snapshot named: %s", snapshotName)
+		snapshotter, err = NewSnapshotter(config.DBRawDSN, stream.logger, snapshotName)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	go func() {
+		defer stream.shutSig.TriggerHasStopped()
+		ctx, done := stream.shutSig.SoftStopCtx(context.Background())
+		defer done()
+		var startLSN LSN
+		if snapshotter != nil {
+			if err = stream.processSnapshot(ctx, snapshotter); err != nil {
 				stream.errors <- fmt.Errorf("failed to process snapshot: %w", err)
 				return
 			}
-			ctx, done := stream.shutSig.SoftStopCtx(context.Background())
-			defer done()
-			if err := stream.startLr(ctx, lsnrestart); err != nil {
-				stream.errors <- fmt.Errorf("failed to start logical replication: %w", err)
+			// TODO: Do we want to ensure all snapshot messages are ack'd before moving
+			// onto the replication stream?
+
+			// Now that the snapshot has been processed, we can copy the replication
+			// slot, represerving the LSN but making it not temporary.
```

Comment:
> but making it not temporary.

What happens if `config.TemporaryReplicationSlot == false`?

Reply:
It's just a rename, technically we could not do this step for temporary slots, but I think temporary slots are just for testing, so I figured simpler code was better. Since why would you want to re-copy the snapshot everytime?